doc/DataPlaneCodeGeneration/Handle-Ambiguity.md

```diff
@@ -0,0 +1,225 @@
+# Handle Ambiguity Between Protocol Method and Convenience Method
+
+## Background
+
+In a DPG library, protocol methods will be generated for a swagger input, and both protocol methods and convenience methods will be generated for a TypeSpec input. We expect to make the convenience method an overload of its protocol method, so that users could be very clear they are representing the same operation. For example,
+
+```C#
+public virtual async Task<Response<Model>> OperationAsync(string requiredId, string optionalId = null, CancellationToken cancellationToken = default) // Convenience method
+{
+    ...
+}
+
+public virtual async Task<Response> OperationAsync(string requiredId, string optionalId = null, RequestContext context = null) // Protocol method
+{
+    ...
+}
+```
+However, this might lead to the compilation error
+```
+Error   CS0121  The call is ambiguous between the following methods or properties: 'SomeClass.OperationAsync(string, CancellationToken)' and 'SomeClass.OperationAsync(string, RequestContext)'
+```
+when calling `OperationAsync("<some string>")`.
+
+This article is to explain how we handle this ambiguity issue and how our solution impacts the generated API.
+
+## Solutions
+
+There are two ways to handle this error:
+
+1. Make all the parameters in **protocol methods** required.
+```C#
+public virtual async Task<Response<Model>> OperationAsync(string requiredId, string optionalId = null, CancellationToken cancellationToken = default) // Convenience method
+{
+    ...
+}
+
+public virtual async Task<Response> OperationAsync(string requiredId, string optionalId, RequestContext context) // Protocol method
+{
+    ...
+}
+```
+2. Append `Value` to the name of the **convenience method**.
+```C#
+public virtual async Task<Response<Model>> OperationValueAsync(string requiredId, string optionalId = null, CancellationToken cancellationToken = default) // Convenience method
+{
+    ...
+}
+
+public virtual async Task<Response> OperationAsync(string requiredId, string optionalId = null, RequestContext context = null) // Protocol method
+{
+    ...
+}
+```
+
+Our strategy is:
+- If the protocol method hasn’t been GAed, the first solution is taken if there exists an ambiguous call.
+- If the protocol method is already GAed, the second solution is taken because we cannot change protocol method anymore.
+
+## Scope of ambiguity we handle
+
+Not all the pairs of protocol method and convenience method will have such problem. Below is a case when there is no ambiguity if users are calling normally by passing a `model` or a `content`. (They could not pass in `null` because it will throw error)
```

Comment:
I think here we should summarize this section in case people would fixate on edge cases and ignoring the problem we are trying to solve such as:
It is impossible that the ambiguities could always be fixed when the two methods are overloads of each other, because of the existence of `default`. We could always use the `default` keyword which is acceptable by any parameter to construct a scenario that the compiler would complain about ambiguities of invocation. Therefore our goal is not solving all cases of the ambiguities, for the following cases, we will allow them to have ambiguities: ...

Reply:
And I see we have two principals in this section, we should emphasize them first and then have examples.